### ISSUE TITLE
Can't set up pipeline component: a factory for 'medspacy_pyrush' already exists

### ISSUE THREAD

**Issue by DSLituiev** [medspacy/medspacy#145]
I've built a fresh container with medspacy==0.2.0.0 and getting this error upon import:

```
Python 3.8.5 (default, Sep  4 2020, 07:30:14) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.19.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import medspacy
2022-09-13 20:08:32.149355: W tensorflow/stream_executor/platform/default/dso_loader.cc:60] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /opt/oracle/instantclient_12_1:/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/amd64/server:
2022-09-13 20:08:32.149411: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
2022-09-13 20:08:33.716570: I tensorflow/compiler/jit/xla_cpu_device.cc:41] Not creating XLA devices, tf_xla_enable_xla_devices not set
2022-09-13 20:08:33.716796: W tensorflow/stream_executor/platform/default/dso_loader.cc:60] Could not load dynamic library 'libcuda.so.1'; dlerror: libcuda.so.1: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /opt/oracle/instantclient_12_1:/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/amd64/server:
2022-09-13 20:08:33.716828: W tensorflow/stream_executor/cuda/cuda_driver.cc:326] failed call to cuInit: UNKNOWN ERROR (303)
2022-09-13 20:08:33.716853: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (run-6320dcbaee1c817bdd9b9d9d-tzwlv): /proc/driver/nvidia/version does not exist
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-1-d39cd9532ab7> in <module>
----> 1 import medspacy

/opt/conda/lib/python3.8/site-packages/medspacy/__init__.py in <module>
      4 from ._extensions import set_extensions, get_extensions, get_doc_extensions, get_span_extensions, get_token_extensions
      5 
----> 6 from . import components
      7 
      8 set_extensions()

/opt/conda/lib/python3.8/site-packages/medspacy/components.py in <module>
     10 }
     11 
---> 12 from .sentence_splitting import PyRuSHSentencizer
     13 from .target_matcher import TargetMatcher
     14 from .context import ConTextComponent

/opt/conda/lib/python3.8/site-packages/medspacy/sentence_splitting.py in <module>
      8 
      9 @Language.factory("medspacy_pyrush")
---> 10 def create_pyrush(nlp, name, pyrush_path=None):
     11     if pyrush_path is None:
     12         pyrush_path = path.join(Path(__file__).resolve().parents[1], "resources", "rush_rules.tsv")

/opt/conda/lib/python3.8/site-packages/spacy/language.py in add_factory(factory_func)
    484                         name=name, func=existing_func, new_func=factory_func
    485                     )
--> 486                     raise ValueError(err)
    487 
    488             arg_names = util.get_arg_names(factory_func)

ValueError: [E004] Can't set up pipeline component: a factory for 'medspacy_pyrush' already exists. Existing factory: <class 'PyRuSH.PyRuSHSentencizer.PyRuSHSentencizer'>. New factory: <function create_pyrush at 0x7f9661a12ee0>
```

Dockerfile:
```
#Base image: quay.io/domino/base:Ubuntu18_DAD_Py3.8_R4.0-20210126

# Gives you Sudo access in container
RUN echo "ubuntu    ALL=NOPASSWD: ALL" >> /etc/sudoers

# Update R Project GPG key
RUN apt-key adv \
  --keyserver keyserver.ubuntu.com \
  --recv-keys E298A3A825C0D65DFD57CBB651716619E084DAB9

# Installs JupyterLab Manager
USER root
RUN jupyter labextension install @jupyter-widgets/jupyterlab-manager@2 && \
	pip install ipywidgets==7.2.1

#USER ubuntu
RUN jupyter nbextension enable --py widgetsnbextension

RUN pip install --upgrade notebook==5.7.5

RUN sudo apt-get clean
RUN rm -f /etc/apt/sources.list.d/sbt.list
RUN apt-get update

RUN apt-get install smbclient -y

#SQL Connection
RUN curl https://packages.microsoft.com/keys/microsoft.asc | sudo apt-key add -
RUN echo "deb [arch=amd64] https://packages.microsoft.com/ubuntu/18.04/prod bionic main" | sudo tee /etc/apt/sources.list.d/mssql-release.list
RUN apt-get update
RUN ACCEPT_EULA=Y apt-get install msodbcsql17

#Python Packages for Pandas + Impala + Google Cloud Platform + PostgreSQL
RUN pip install sqlalchemy
RUN pip install pandasql
RUN pip install impyla
RUN pip install thrift-sasl
RUN pip install pandas
RUN pip install --upgrade google-cloud-bigquery
#Updated versions of ca certificates typically fixes release file errors
RUN sudo apt install ca-certificates
# Upgrade postgresql repo list to https
RUN sed -i 's/http/https/g' /etc/apt/sources.list.d/pgdg.list
RUN sudo apt-get update
RUN sudo apt-get remove libpq5 -y
RUN sudo apt-get install libpq-dev -y
RUN pip install psycopg2
RUN pip install PyGreSQL
RUN pip install pysftp
RUN pip install boto3
RUN pip install janitor
RUN pip install pyjanitor

#R Packages instalation start - Priyankar
USER root

RUN sudo apt-mark hold curl
RUN sudo apt-get update
RUN sudo apt --fix-broken install -y
RUN sudo apt-get install -y libv8-dev


RUN pip install importlib_metadata==3.8.2 && \
    numpy==1.15.0 && \
    pip install spacy==3.2.4 && \
    pip install medspacy==0.2.0.0 && \
    pip install medspacy-quickumls==2.4.1 && \
    pip install scispacy==0.5.0 && \
    pip install https://s3-us-west-2.amazonaws.com/ai2-s2-scispacy/releases/v0.5.1/en_core_sci_lg-0.5.1.tar.gz
```

**Comment by DSLituiev** [comment]
This seems to be caused as the error says by this class from PyRush:

`PyRuSH/PyRuSHSentencizer.py:25`

```python
@Language.factory("medspacy_pyrush")
class PyRuSHSentencizer(Sentencizer):
    def __init__(self, nlp: Language, name: str = "medspacy_pyrush", rules_path: str = '', max_repeat: int = 50,
                 auto_fix_gaps: bool = True) -> Sentencizer:
              # ...
```

Thus medspacy is trying to override a registered class that it imports.

Is it a recent compatibility issue?
> Start from version 1.0.3, PyRuSH adds Spacy compatible Sentencizer component: PyRuSHSentencizer.
https://github.com/jianlins/PyRuSH#spacy-componentized-pyrush

**Comment by abchapman93** [comment]
Hi @DSLituiev ,
Could you verify which versions of medspaCy PyRuSH you have installed? I found this issue the other day and updated medspaCy to require a lower version of PyRuSH while we wait for the original issue to be fixed.

I believe that installing PyRuSH 1.0.3.6 should do the trick.

**Comment by DSLituiev** [comment]
Hi @abchapman93 I am using `medspacy==0.2.0.0` as you can see from the Dockerfile.

I tried force installing  `PyRuSH==1.0.2` (as PyRuSH announces Spacy compatible Sentencizer component from version 1.0.3 onwards), but medspacy kicks it up to a later version.

When I go to `medspacy==0.2.0.0` tag on github and look into the [`setup.py`](https://github.com/medspacy/medspacy/blob/0.2.0.0/setup.py#L58), I see that it requires  `"PyRuSH>=1.0.3.5"`. So looks like this is a bug that can be monkeypatched, but probably better to fix it in the distribution

**Comment by abchapman93** [comment]
Try updating to medspacy 0.2.0.1, that version has the updated version for PyRuSH.

**Comment by DSLituiev** [comment]
Thank you @abchapman93. Your previous recommendation worked:
```
RUN pip install importlib_metadata==3.8.2 && \
    numpy==1.15.0 && \
    pip install PyRuSH==1.0.3.6 && \
    pip install spacy==3.2.4 && \
    pip install medspacy==0.2.0.0 && \
    pip install medspacy-quickumls==2.4.1 && \
    pip install scispacy==0.5.0 && \
    pip install https://s3-us-west-2.amazonaws.com/ai2-s2-scispacy/releases/v0.5.1/en_core_sci_lg-0.5.1.tar.gz
```

**Comment by hannapethani** [comment]
@DSLituiev 

Interesting you managed to install spacy 3.2.4. with medspacy 0.2.0.0 and with 0.2.0.1. I received an error and failed install using python 3.8.

```
The conflict is caused by:
The user requested spacy==3.2.4
pyrush 1.0.3.6 depends on spacy>=2.2.2
medspacy 0.2.0.0 depends on spacy<3.2.0 and >=3.1.3

Pip subprocess error:
ERROR: Cannot install -r condaenv.jbeviudf.requirements.txt (line 1), -r 
and spacy==3.2.4 because these package versions have conflicting dependencies.
```

**Comment by burgersmoke** [comment]
Since @abchapman93  resolved this issue above with medspacy 0.2.0.1, and since this is also resolved with medspacy 1.0.0 released yesterday, I am inclined to close this issue.  If anyone feels otherwise, please open a new ticket, but the initial issue opened has been resolved (i.e., "Can't set up pipeline component: a factory for 'medspacy_pyrush' already exist")
